### ISSUE TITLE
Issues making lodash available globally

### ISSUE THREAD

**Issue by MikeJoyce19** [jspm/registry#431]
I have an internal property that is not jspm compliant that has a global dependency on lodash and jquery.  It appears that jquery is being made available globally however lodash is not.  Here is the override in my package.json file:

```
"overrides": {
      "npm:mypackage@2.0.1641": {
        "dependencies": {
          "jquery": ">= 1.x",
          "lodash": ">= 2.x",
          "d3": ">= 3.x",
          "highstock-release": ">= 2.0.4"
        },
        "shim": {
          "mypackage": {
            "deps": [
              "jquery",
              "lodash",
              "d3"
            ]
          }
        }
      }
    }

```

This is resulting in mypackage and jquery to be available globally but lodash is not. I get a reference error in my package's file saying that ```"_" can't be found ``` . Why is jquery available globally but lodash isnt?

**Comment by guybedford** [comment]
Just answered this in chat room...

> Yes the version of lodash we use doesn't define a global
> I'd suggest making sure you import it in the package loading it via require('lodash') or a wrapper that sets the global alternatively this can also be done via configuration -

```javascript
System.config({
  meta: {
    'npm:mike@2.0.1641/mike.js': { globals: { lodash: 'lodash' } }
  }
});
```

> note that the above will only work in the beta release of jspm though as globals is a new feature in SystemJS exactly for this scenario

**Comment by maximnaidenov** [comment]
I just met the same problem that npm:lodash does not define global. But github:lodash/lodash does export '_' so I just installed it and replaced manually the dependency to npm:lodash with github:lodash.

**Comment by maximnaidenov** [comment]
Just for someone facing the same issue. With jspm@16 ( that uses SystemJS@0.17) it is now possible to shim the global dependency that restangular has to lodash with this configuration in config.js:
````
meta: {
    'github:mgonto/restangular@1.5.1/restangular.js': {
      globals: {
        "_": "lodash"
      }
    }
````

**Comment by vigie** [comment]
Is there some configuration we can check in as an override for restangular and similar global modules so that everyone using a global having a dependency on lodash does not need to add these meta blocks? For example, should such global modules revert to using the github version of lodash as suggested above? Or is adding the meta config to every project seen as the correct way of doing things? @guybedford Seems like the ability to configure globals for a module as a jspm override is needed? Thanks.

**Comment by guybedford** [comment]
The globals property will only be supported in overrides from jspm 0.17. Because lodash is expected to be a global, it's not just a shim override here.

**Comment by ovidiom** [comment]
I've defined the globals in the same way as @maximnaidenov above which works fine but I've noticed that the "_" variable is not only available during the execution of the Restangular global module but it remains available as a global throughout the exeuction of my application.

Is this the intended behaviour? From the docs I understood that the global would only be temporarily set during the execution of the global module:
https://github.com/systemjs/systemjs/blob/master/docs/module-formats.md#custom-globals

**Comment by guybedford** [comment]
@ovidiom good catch - that's a bug, resolved in https://github.com/systemjs/systemjs/commit/5b8bfbf6cc0dacaf65968e00f0f39c51d25daf9d.

**Comment by ovidiom** [comment]
Thanks for the feedback and for fixing is so quickly!

**Comment by dave0783** [comment]
So is this fixed or am I doing something wrong?

I am adding the following to `config.js` as suggested:

```
meta: {
    'github:mgonto/restangular@1.5.1/restangular.js': {
      globals: {
        "_": "lodash"
      }
}
```

All this seems to do is change the error from `ReferenceError: _ is not defined` to `TypeError: Cannot read property 'isUndefined' of undefined`

Using `JSPM v0.16.19` and `SystemJS v0.19.9`

**Comment by guybedford** [comment]
@dave0783 I guess the first thing to test here is that `System.import('lodash')` is defined correctly. If so then note that the `globals` metadata only works for modules that are CJS or global format. You may need to add a `"format": "global"` meta there along with `globals`.

**Comment by circlingthesun** [comment]
So it appears that given @dave0783's config, `_` is defined when the module is loaded. However, when provider is initialised by angular, `_` is no longer defined.

I added some console.logs in the source:

```js
/* */ 
"format global";
"deps angular";
"deps lodash";
console.log("_ is defined?", !!_);
/**
 * Restful Resources service for AngularJS apps
 * @version v1.4.0 - 2015-04-03 * @link https://github.com/mgonto/restangular
 * @author Martin Gontovnikas <martin@gon.to>
 * @license MIT License, http://www.opensource.org/licenses/MIT
 */(function() {

var restangular = angular.module('restangular', []);

restangular.provider('Restangular', function() {
    console.log("_ is defined?", !!_);
....
```

And the output is as follows:
```
_ is defined? true
_ is defined? false
```

I'm running SystemJS v0.19.18.

**Comment by guybedford** [comment]
@circlingthesun try using the `globals` configuration instead of dependency shimming here rather.

(this is effectively a globals meta or "globals._ lodash"; meta syntax in the source)

**Comment by dave0783** [comment]
>@dave0783 I guess the first thing to test here is that System.import('lodash') is defined correctly. If so then note that the globals metadata only works for modules that are CJS or global format. You may need to add a "format": "global" meta there along with globals.

I tried this with no success:

```
meta: {
    'github:mgonto/restangular@1.5.1/restangular.js': {
      globals: {
          "_": "npm:lodash@3.10.1"
      },
      "format": "global"
    }
  }
```
I am not sure if thats what you meant or not.

>@circlingthesun try using the globals configuration instead of dependency shimming here rather.
(this is effectively a globals meta or "globals._ lodash"; meta syntax in the source)

Where would we put `"globals._lodash";`?

**Comment by guybedford** [comment]
@dave0783 that's exactly right. The above should definitely work if `import _ from 'npm:lodash@3.10.1'` works.

**Comment by circlingthesun** [comment]
@guybedford, I'm using the same globals config as @dave0783 is referencing above. Some how `_` is not captured by the module's closure or it is unset after executing the module body. I've managed to work around this issue by using `github:lodash/lodash` instead of `npm:lodash` as @maximnaidenov pointed out.

**Comment by guybedford** [comment]
I just tested this out and can confirm it is definitely working fine. It sounds like the issue is in properly applying the meta configuration. An override PR is welcome here.

**Comment by dave0783** [comment]
@guybedford can you post how you have this configured and how you are importing restangular and lodash? No matter what I do, as soon as I try to configure restangular, it complains about missing functions of lodash

**Comment by guybedford** [comment]
@dave0783 sure, `jspm install restangular=github:mgonto/restangular -o "{shim:{'restangular':{deps:['angular', 'lodash']}}}"` and `System.import('restangular')`.

**Comment by noelmace** [comment]
Thanks to your posts (especially the last onex from @dave0783 and @vigie), I finally managed to make it work.

Here is my configuration, if you need it.

package.json
```json
"dependencies": {
      "angular": "github:angular/bower-angular@^1.5.3",
      "lodash": "github:lodash/lodash@3.10.1",
      "restangular": "github:mgonto/restangular@^1.5.2"
    },
"overrides": {
      "github:mgonto/restangular@1.5.2": {
        "shim": {
          "restangular": {
            "deps": [
              "angular",
              "lodash"
            ]
          }
        }
      }
    }
```

system.config.js
```javascript
"github:mgonto/restangular@1.5.2/restangular.js": {
      "format": "global",
      "globals": {
        "_": "lodash"
      }
    }
```

**Comment by ciel** [comment]
Has this changed since this topic was made? I spent hours upon hours trying to make this work with a `markdown` library, like this ...

```
"lib/markdown": {
    "format": "global",
    "globals": { "toMarkdown": "toMarkdown" }
}
```

but it turned out, I needed to do **global** , not global**s**, and define an **exports** property, like this.

```
"lib/markdown": {
    "exports": "toMarkdown",
    "global": { "toMarkdown": "toMarkdown" }
}
```

I couldn't find much in the way of docs that talked about this, and it really took several hours from me. It would sure be swell if the jspm docs could touch on this even briefly. Thanks again for everything you guys do! 

Do you accept pull requests on the docs? Could I write instructions for this myself and offer them up?

**Comment by circlingthesun** [comment]
With restangular I've noticed that the UMD wrapper gets removed from restangular.js. How can I avoid that?
